pywbem_mock/_defaultinstanceprovider.py

```diff
@@ -0,0 +1,527 @@
+#
+# (C) Copyright 2018 InovaDevelopment.com
+#
+# This library is free software; you can redistribute it and/or
+# modify it under the terms of the GNU Lesser General Public
+# License as published by the Free Software Foundation; either
+# version 2.1 of the License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
+#
+# Author: Karl  Schopmeyer <inovadevelopment.com>
+#
+
+"""
+This module adds support for user-defined providers.  User defined
+providers may be created by the use of pywbem_mock to extend the capability
+of the mocker for special processing for certain classes to be mocked.
+
+User defined providers may be created for specific classnames to implement the
+CreateInstance and DeleteInstance methods and override the default corresponding
+methods in DefaultInstanceProvider.
+
+TODO: Defined the rules for user-defined providers.
+
+This module contains two classes:
+
+1. ProviderDispatcher - Routes the calls for request methods that allow
+user-defined providers either to the default method in DefaultInstanceProvider
+or to a user defined method in the provider registry.
+
+2. DefaultInstanceProvider - The default implementation for the request
+responders that may include user defined providers.  This is the method that
+will be called if there is no provider registered for a namespace and classname.
+
+"""
+
+from __future__ import absolute_import, print_function
+
+import six
+
+
+from pywbem import CIMInstance, CIMInstanceName, CIMError, \
+    CIM_ERR_NOT_FOUND, CIM_ERR_INVALID_PARAMETER, CIM_ERR_INVALID_CLASS, \
+    CIM_ERR_ALREADY_EXISTS
+
+from pywbem._utils import _format
+
+from ._baseprovider import BaseProvider
+
+from ._baseprovider import INSTANCE_RETRIEVE_LOCAL_ONLY
+
+# None of the request method names conform since they are camel case
+# pylint: disable=invalid-name
+
+
+class ProviderDispatcher(BaseProvider):
```

Comment:
It seems strange that a provider dispatcher should itself be a provider.

Reply:
I know.   However, with your earlier comments we have 2 things to do before we hand over to the provider.

1. Decide which provider (default or user)

2. Handle the validation processing (your comments).  To do this, the container has to include access to the methods that actually validate namespace, validate class exists, etc. That code, in turn requires access to the cimrepository. 

One Alternative I was thinking about was to create new methods in the DefaultInstanceProvider that would be something like CreateInstanceDispatcher and DeleteInstanceDispatcher,  That would eliminate the ProviderDispatcher completely. Each of these methods would 1. Validate request parameters, 2. Determine if the default is to be called or the user provider.  

**DISCUSSION**

Reply:
I think the provider dispatcher should also have a reference to the repository.

Reply:
It does now.